File: packages/shared/src/billing/reports/trend.ts
import type { WorklogRow, ContractRow } from '../types.js';
import { bucketKey, type Granularity } from './buckets.js';

// Post-#108 the app is CZK-only (rate_currency dropped); every earned amount is CZK.
const isCzkEarned = (r: WorklogRow) => r.earnedAmount != null;
const cmpStr = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);

export interface TrendBucket {
  bucket: string;
  minutes: number;
  earnedCzk: number;
}

export function trendSeries(
  rows: WorklogRow[],
  opts: { from: string; to: string; granularity: Granularity; projectId?: number },
): TrendBucket[] {
  const { from, to, granularity, projectId } = opts;
  const map = new Map<string, TrendBucket>();
  for (const r of rows) {
    if (r.workDate < from || r.workDate > to) continue;
    if (projectId !== undefined && r.projectId !== projectId) continue;
    const key = bucketKey(r.workDate, granularity);
    const cur = map.get(key) ?? { bucket: key, minutes: 0, earnedCzk: 0 };
    cur.minutes += r.effectiveMinutes;
    if (isCzkEarned(r)) cur.earnedCzk += r.earnedAmount!;
    map.set(key, cur);
  }
  return [...map.values()].sort((a, b) => cmpStr(a.bucket, b.bucket));
}

export interface RateMarker {
  effectiveFrom: string;
  rateType: 'hourly' | 'daily';
  rateAmount: number;
}

export function rateChangeMarkers(
  contracts: ContractRow[],
  opts: { from: string; to: string; projectId?: number },
): RateMarker[] {
  const { from, to, projectId } = opts;
  if (projectId === undefined) return [];
  const ordered = contracts
    .filter((c) => c.projectId === projectId)
    .sort((a, b) => cmpStr(a.effectiveFrom, b.effectiveFrom));
  return ordered
    .slice(1) // rank > 1 — skip the earliest contract (not a "change")
    .filter((c) => c.effectiveFrom >= from && c.effectiveFrom <= to)
    .map((c) => ({
      effectiveFrom: c.effectiveFrom,
      rateType: c.rateType,
      rateAmount: c.rateAmount,
    }));
}

File: swift/WatchtowerCore/Sources/WatchtowerCore/Billing/reports/Trend.swift
import Foundation

/// Per-bucket trend point — mirrors `packages/shared/src/billing/reports/trend.ts`.
public struct TrendBucket: Equatable, Sendable {
    public let bucket: String
    public let minutes: Double
    public let earnedCzk: Double

    public init(bucket: String, minutes: Double, earnedCzk: Double) {
        self.bucket = bucket
        self.minutes = minutes
        self.earnedCzk = earnedCzk
    }
}

/// A contract effective-date change worth marking on a trend chart.
public struct RateMarker: Equatable, Sendable {
    public let effectiveFrom: String
    public let rateType: String
    public let rateAmount: Double

    public init(effectiveFrom: String, rateType: String, rateAmount: Double) {
        self.effectiveFrom = effectiveFrom
        self.rateType = rateType
        self.rateAmount = rateAmount
    }
}

/// Buckets `rows` by `granularity` within `[from, to]`, optionally scoped to
/// `projectId`. Sums `effectiveMinutes` always; sums `earnedAmount` only for
/// rows where it is non-nil. Result is sorted ascending by bucket key.
public func trendSeries(
    _ rows: [WorklogRow], from: String, to: String, granularity: Granularity, projectId: Int?
) -> [TrendBucket] {
    var buckets: [String: TrendBucket] = [:]
    for r in rows {
        if r.workDate < from || r.workDate > to { continue }
        if let projectId, r.projectId != projectId { continue }
        let key = bucketKey(r.workDate, granularity)
        let cur = buckets[key] ?? TrendBucket(bucket: key, minutes: 0, earnedCzk: 0)
        let earned = cur.earnedCzk + (r.earnedAmount ?? 0)
        buckets[key] = TrendBucket(bucket: key, minutes: cur.minutes + r.effectiveMinutes, earnedCzk: earned)
    }
    return buckets.values.sorted { $0.bucket < $1.bucket }
}

/// Rate-change markers for `projectId`'s contracts within `[from, to]`. The
/// earliest contract (by `effectiveFrom`) is dropped — it's the starting
/// rate, not a "change". Returns `[]` when `projectId` is nil.
public func rateChangeMarkers(
    _ contracts: [ContractRow], from: String, to: String, projectId: Int?
) -> [RateMarker] {
    guard let projectId else { return [] }
    let ordered = contracts
        .filter { $0.projectId == projectId }
        .sorted { $0.effectiveFrom < $1.effectiveFrom }
    return ordered
        .dropFirst()
        .filter { $0.effectiveFrom >= from && $0.effectiveFrom <= to }
        .map { RateMarker(effectiveFrom: $0.effectiveFrom, rateType: $0.rateType, rateAmount: $0.rateAmount) }
}
